File: cli-rust/src/context/localconfig.rs
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::io;
use std::path::PathBuf;

#[derive(Serialize, Deserialize)]
pub struct Contents {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
}

pub struct Config {
    pub path: PathBuf,
    pub contents: Contents,
}

impl Config {
    fn read(path: &PathBuf) -> io::Result<Contents> {
        let file = File::open(path)?;
        let result = serde_json::from_reader(file)?;
        Ok(result)
    }

    pub fn load(path: &PathBuf) -> io::Result<Self> {
        match Config::read(path) {
            Ok(contents) => Ok(Config {
                path: path.clone(),
                contents,
            }),
            Err(_) => {
                let mut config = Config {
                    path: path.clone(),
                    contents: Contents {
                        token: None,
                        endpoint: None,
                    },
                };
                config.save()?;
                return Ok(config);
            }
        }
    }

    pub fn save(&mut self) -> io::Result<()> {
        let file = File::create(&self.path)?;
        Ok(serde_json::to_writer(file, &self.contents)?)
    }
}

File: cli-go/localconfig/localconfig.go
package localconfig

import (
	"encoding/json"
	"io/ioutil"
)

type Config struct {
	Token    string
	Endpoint string
	path     string
}

func Init(path string) (*Config, error) {
	var config Config
	config.path = path

	err := config.Load()
	if err != nil {
		err = config.Save()
		if err != nil {
			return &config, err
		}
	}

	return &config, nil
}

func (config *Config) Load() error {
	contents, err := ioutil.ReadFile(config.path)
	if err != nil {
		return err
	}

	err = json.Unmarshal(contents, config)
	if err != nil {
		return err
	}
	return nil
}

func (config *Config) Save() error {
	contents, err := json.MarshalIndent(config, "", "    ")
	if err != nil {
		return err
	}

	err = ioutil.WriteFile(config.path, contents, 0o644)
	if err != nil {
		return err
	}

	return nil
}
